config_diff.txt
// Taken from https://github.com/djpohly/dwl/issues/466	      |	/* Taken from https://github.com/djpohly/dwl/issues/466 */
#define COLOR(hex)    { ((hex >> 24) & 0xFF) / 255.0f, \	#define COLOR(hex)    { ((hex >> 24) & 0xFF) / 255.0f, \
                        ((hex >> 16) & 0xFF) / 255.0f, \	                        ((hex >> 16) & 0xFF) / 255.0f, \
                        ((hex >> 8) & 0xFF) / 255.0f, \		                        ((hex >> 8) & 0xFF) / 255.0f, \
                        (hex & 0xFF) / 255.0f }			                        (hex & 0xFF) / 255.0f }
/* appearance */						/* appearance */
static const int sloppyfocus               = 1;  /* focus fol	static const int sloppyfocus               = 1;  /* focus fol
static const int bypass_surface_visibility = 0;  /* 1 means i	static const int bypass_surface_visibility = 0;  /* 1 means i
static const unsigned int borderpx         = 1;  /* border pi	static const unsigned int borderpx         = 1;  /* border pi
static const int showbar                   = 1; /* 0 means no	static const int showbar                   = 1; /* 0 means no
static const int topbar                    = 1; /* 0 means bo	static const int topbar                    = 1; /* 0 means bo
static const char *fonts[]                 = {"monospace:size	static const char *fonts[]                 = {"monospace:size
static const float rootcolor[]             = COLOR(0x000000ff	static const float rootcolor[]             = COLOR(0x000000ff
/* This conforms to the xdg-protocol. Set the alpha to zero t	/* This conforms to the xdg-protocol. Set the alpha to zero t
static const float fullscreen_bg[]         = {0.0f, 0.0f, 0.0	static const float fullscreen_bg[]         = {0.0f, 0.0f, 0.0
static uint32_t colors[][3]                = {			static uint32_t colors[][3]                = {
	/*               fg          bg          border    */		/*               fg          bg          border    */
	[SchemeNorm] = { 0xbbbbbbff, 0x222222ff, 0x444444ff }		[SchemeNorm] = { 0xbbbbbbff, 0x222222ff, 0x444444ff }
	[SchemeSel]  = { 0xeeeeeeff, 0x005577ff, 0x005577ff }		[SchemeSel]  = { 0xeeeeeeff, 0x005577ff, 0x005577ff }
	[SchemeUrg]  = { 0,          0,          0x770000ff }		[SchemeUrg]  = { 0,          0,          0x770000ff }
};								};

/* tagging */							/* tagging */
static char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8	static char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8

/* logging */							/* logging */
static int log_level = WLR_ERROR;				static int log_level = WLR_ERROR;

							      >	/* Autostart */
							      >	static const char *const autostart[] = {
							      >	        "wbg", "/path/to/your/image", NULL,
							      >	        NULL /* terminate */
							      >	};
							      >
static const Rule rules[] = {					static const Rule rules[] = {
	/* app_id             title       tags mask     isflo		/* app_id             title       tags mask     isflo
	{ "Gimp_EXAMPLE",     NULL,       0,            1,   		{ "Gimp_EXAMPLE",     NULL,       0,            1,   
	{ "firefox_EXAMPLE",  NULL,       1 << 8,       0,   		{ "firefox_EXAMPLE",  NULL,       1 << 8,       0,   
    /* default/example rule: can be changed but cannot be eli	    /* default/example rule: can be changed but cannot be eli
};								};

/* layout(s) */							/* layout(s) */
static const Layout layouts[] = {				static const Layout layouts[] = {
	/* symbol     arrange function */				/* symbol     arrange function */
	{ "[]=",      tile },						{ "[]=",      tile },
	{ "><>",      NULL },    /* no layout function means 		{ "><>",      NULL },    /* no layout function means 
	{ "[M]",      monocle },					{ "[M]",      monocle },
							      >		{ "[E]",      deck },
};								};

/* monitors */							/* monitors */
/* (x=-1, y=-1) is reserved as an "autoconfigure" monitor pos	/* (x=-1, y=-1) is reserved as an "autoconfigure" monitor pos
 * WARNING: negative values other than (-1, -1) cause problem	 * WARNING: negative values other than (-1, -1) cause problem
 * https://gitlab.freedesktop.org/xorg/xserver/-/issues/899 *	 * https://gitlab.freedesktop.org/xorg/xserver/-/issues/899 *
static const MonitorRule monrules[] = {				static const MonitorRule monrules[] = {
   /* name        mfact  nmaster scale layout       rotate/re	   /* name        mfact  nmaster scale layout       rotate/re
    * example of a HiDPI laptop monitor:			    * example of a HiDPI laptop monitor:
    { "eDP-1",    0.5f,  1,      2,    &layouts[0], WL_OUTPUT	    { "eDP-1",    0.5f,  1,      2,    &layouts[0], WL_OUTPUT
    { "HDMI-A-1", 0.5f,  4,      1,    &layouts[0], WL_OUTPUT |		{ NULL,       0.55f, 1,      1,    &layouts[0], WL_OU
    { "DP-2",     0.5f,  4,      1,    &layouts[0], WL_OUTPUT |		/* default monitor rule: can be changed but cannot be
							      >	    /*{ "HDMI-A-1", 0.5f,  4,      1,    &layouts[0], WL_OUTP
							      >	    { "DP-2",     0.5f,  4,      1,    &layouts[0], WL_OUTPUT
};								};

/* keyboard */							/* keyboard */
static const struct xkb_rule_names xkb_rules = {		static const struct xkb_rule_names xkb_rules = {
	/* can specify fields: rules, model, layout, variant,		/* can specify fields: rules, model, layout, variant,
	/* example:							/* example:
	.options = "ctrl:nocaps",					.options = "ctrl:nocaps",
	*/								*/
	.layout = "br",							.layout = "br",
	.variant = "abnt2",						.variant = "abnt2",
	.options = NULL,						.options = NULL,
};								};

static const int repeat_rate = 25;				static const int repeat_rate = 25;
static const int repeat_delay = 600;				static const int repeat_delay = 600;

/* Start of user-made configuration */			      <
							      <
/* Autostart */						      <
static const char *const autostart[] = {		      <
       /* "swaybg", "-o", "HDMI-A-1", "-i", "/home/rafic/Pict <
        "swaybg", "-o", "DP-2", "-i", "/home/rafic/Pictures/W <
        "/home/rafic/.config/dwl/autostart.sh", NULL,	      <
        NULL /* terminate */				      <
};							      <
							      <
/* End of user-made configuration */			      <
							      <
/* Trackpad */							/* Trackpad */
static const int tap_to_click = 1;				static const int tap_to_click = 1;
static const int tap_and_drag = 1;				static const int tap_and_drag = 1;
static const int drag_lock = 1;					static const int drag_lock = 1;
static const int natural_scrolling = 0;				static const int natural_scrolling = 0;
static const int disable_while_typing = 1;			static const int disable_while_typing = 1;
static const int left_handed = 0;				static const int left_handed = 0;
static const int middle_button_emulation = 0;			static const int middle_button_emulation = 0;
/* You can choose between:					/* You can choose between:
LIBINPUT_CONFIG_SCROLL_NO_SCROLL				LIBINPUT_CONFIG_SCROLL_NO_SCROLL
LIBINPUT_CONFIG_SCROLL_2FG					LIBINPUT_CONFIG_SCROLL_2FG
LIBINPUT_CONFIG_SCROLL_EDGE					LIBINPUT_CONFIG_SCROLL_EDGE
LIBINPUT_CONFIG_SCROLL_ON_BUTTON_DOWN				LIBINPUT_CONFIG_SCROLL_ON_BUTTON_DOWN
*/								*/
static const enum libinput_config_scroll_method scroll_method	static const enum libinput_config_scroll_method scroll_method

/* You can choose between:					/* You can choose between:
LIBINPUT_CONFIG_CLICK_METHOD_NONE				LIBINPUT_CONFIG_CLICK_METHOD_NONE
LIBINPUT_CONFIG_CLICK_METHOD_BUTTON_AREAS			LIBINPUT_CONFIG_CLICK_METHOD_BUTTON_AREAS
LIBINPUT_CONFIG_CLICK_METHOD_CLICKFINGER			LIBINPUT_CONFIG_CLICK_METHOD_CLICKFINGER
*/								*/
static const enum libinput_config_click_method click_method =	static const enum libinput_config_click_method click_method =

/* You can choose between:					/* You can choose between:
LIBINPUT_CONFIG_SEND_EVENTS_ENABLED				LIBINPUT_CONFIG_SEND_EVENTS_ENABLED
LIBINPUT_CONFIG_SEND_EVENTS_DISABLED				LIBINPUT_CONFIG_SEND_EVENTS_DISABLED
LIBINPUT_CONFIG_SEND_EVENTS_DISABLED_ON_EXTERNAL_MOUSE		LIBINPUT_CONFIG_SEND_EVENTS_DISABLED_ON_EXTERNAL_MOUSE
*/								*/
static const uint32_t send_events_mode = LIBINPUT_CONFIG_SEND	static const uint32_t send_events_mode = LIBINPUT_CONFIG_SEND

/* You can choose between:					/* You can choose between:
LIBINPUT_CONFIG_ACCEL_PROFILE_FLAT				LIBINPUT_CONFIG_ACCEL_PROFILE_FLAT
LIBINPUT_CONFIG_ACCEL_PROFILE_ADAPTIVE				LIBINPUT_CONFIG_ACCEL_PROFILE_ADAPTIVE
*/								*/
static const enum libinput_config_accel_profile accel_profile	static const enum libinput_config_accel_profile accel_profile
static const double accel_speed = 0.0;				static const double accel_speed = 0.0;

/* You can choose between:					/* You can choose between:
LIBINPUT_CONFIG_TAP_MAP_LRM -- 1/2/3 finger tap maps to left/	LIBINPUT_CONFIG_TAP_MAP_LRM -- 1/2/3 finger tap maps to left/
LIBINPUT_CONFIG_TAP_MAP_LMR -- 1/2/3 finger tap maps to left/	LIBINPUT_CONFIG_TAP_MAP_LMR -- 1/2/3 finger tap maps to left/
*/								*/
static const enum libinput_config_tap_button_map button_map =	static const enum libinput_config_tap_button_map button_map =

/* If you want to use the windows key for MODKEY, use WLR_MOD	/* If you want to use the windows key for MODKEY, use WLR_MOD
#define MODKEY WLR_MODIFIER_LOGO				#define MODKEY WLR_MODIFIER_LOGO

#define TAGKEYS(KEY,SKEY,TAG) \					#define TAGKEYS(KEY,SKEY,TAG) \
	{ MODKEY,                    KEY,            view,   		{ MODKEY,                    KEY,            view,   
	{ MODKEY|WLR_MODIFIER_CTRL,  KEY,            togglevi		{ MODKEY|WLR_MODIFIER_CTRL,  KEY,            togglevi
	{ MODKEY|WLR_MODIFIER_SHIFT, SKEY,           tag,    		{ MODKEY|WLR_MODIFIER_SHIFT, SKEY,           tag,    
	{ MODKEY|WLR_MODIFIER_CTRL|WLR_MODIFIER_SHIFT,SKEY,to		{ MODKEY|WLR_MODIFIER_CTRL|WLR_MODIFIER_SHIFT,SKEY,to

/* helper for spawning shell commands in the pre dwm-5.0 fash	/* helper for spawning shell commands in the pre dwm-5.0 fash
#define SHCMD(cmd) { .v = (const char*[]){ "/bin/sh", "-c", c	#define SHCMD(cmd) { .v = (const char*[]){ "/bin/sh", "-c", c

/* commands */							/* commands */
static const char *termcmd[] = { "kitty", NULL };	      |	static const char *termcmd[] = { "foot", NULL };
static const char *menucmd[] = { "fuzzel", NULL };	      |	static const char *menucmd[] = { "wmenu-run", NULL };
							      <
/* Screenshot command config start */			      <
static const char *screenshot_full[]={"/home/rafic/.local/bin <
static const char *screenshot_area[]={"/home/rafic/.local/bin <
static const char *screenshot_clipboard[]={"/home/rafic/.loca <
/* Screenshot command config end */			      <

static const Key keys[] = {					static const Key keys[] = {
	/* Note that Shift changes certain key codes: 2 -> at		/* Note that Shift changes certain key codes: 2 -> at
	/* modifier                  key                  fun		/* modifier                  key                  fun
	{ MODKEY,                    XKB_KEY_d,           spa		{ MODKEY,                    XKB_KEY_d,           spa
	{ MODKEY, 		     XKB_KEY_t,      	  spa |		{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_t,      spawn,  
	{ MODKEY,                    XKB_KEY_b,           tog		{ MODKEY,                    XKB_KEY_b,           tog
	{ MODKEY,                    XKB_KEY_j,           foc		{ MODKEY,                    XKB_KEY_j,           foc
	{ MODKEY,                    XKB_KEY_k,           foc		{ MODKEY,                    XKB_KEY_k,           foc
	{ MODKEY,                    XKB_KEY_i,           inc		{ MODKEY,                    XKB_KEY_i,           inc
	{ MODKEY,                    XKB_KEY_p,           inc		{ MODKEY,                    XKB_KEY_p,           inc
	{ MODKEY,                    XKB_KEY_h,           set		{ MODKEY,                    XKB_KEY_h,           set
	{ MODKEY,                    XKB_KEY_l,           set		{ MODKEY,                    XKB_KEY_l,           set
	{ MODKEY,                    XKB_KEY_Return,      zoo		{ MODKEY,                    XKB_KEY_Return,      zoo
	{ MODKEY,                    XKB_KEY_Tab,         vie		{ MODKEY,                    XKB_KEY_Tab,         vie
	{ MODKEY, 		     XKB_KEY_q,           kil |		{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_c,           kil
	{ WLR_MODIFIER_CTRL|WLR_MODIFIER_ALT,    XKB_KEY_t,   |		{ MODKEY,                    XKB_KEY_t,           set
	{ WLR_MODIFIER_CTRL|WLR_MODIFIER_ALT,    XKB_KEY_f,   |		{ MODKEY,                    XKB_KEY_f,           set
	{ WLR_MODIFIER_CTRL|WLR_MODIFIER_ALT,    XKB_KEY_m,   |		{ MODKEY,                    XKB_KEY_m,           set
	{ MODKEY,                    XKB_KEY_space,       set		{ MODKEY,                    XKB_KEY_space,       set
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_space,       tog		{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_space,       tog
	{ MODKEY,                    XKB_KEY_e,           tog		{ MODKEY,                    XKB_KEY_e,           tog
	{ MODKEY,                    XKB_KEY_0,           vie		{ MODKEY,                    XKB_KEY_0,           vie
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_parenright,  tag		{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_parenright,  tag
	{ MODKEY,                    XKB_KEY_comma,       foc		{ MODKEY,                    XKB_KEY_comma,       foc
	{ MODKEY,                    XKB_KEY_period,      foc		{ MODKEY,                    XKB_KEY_period,      foc
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_less,        tag		{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_less,        tag
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_greater,     tag		{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_greater,     tag
	/* Start of user-made hotkeys */		      <
	{ MODKEY,		     		 XKB_KEY_s,   <
	{ MODKEY|WLR_MODIFIER_ALT,   XKB_KEY_s,           spa <
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_s,           spa <
	/* End of user-made hotkeys */			      <
	TAGKEYS(          XKB_KEY_1, XKB_KEY_exclam,         		TAGKEYS(          XKB_KEY_1, XKB_KEY_exclam,         
	TAGKEYS(          XKB_KEY_2, XKB_KEY_at,             		TAGKEYS(          XKB_KEY_2, XKB_KEY_at,             
	TAGKEYS(          XKB_KEY_3, XKB_KEY_numbersign,     		TAGKEYS(          XKB_KEY_3, XKB_KEY_numbersign,     
	TAGKEYS(          XKB_KEY_4, XKB_KEY_dollar,         		TAGKEYS(          XKB_KEY_4, XKB_KEY_dollar,         
	TAGKEYS(          XKB_KEY_5, XKB_KEY_percent,        		TAGKEYS(          XKB_KEY_5, XKB_KEY_percent,        
	TAGKEYS(          XKB_KEY_6, XKB_KEY_asciicircum,    		TAGKEYS(          XKB_KEY_6, XKB_KEY_asciicircum,    
	TAGKEYS(          XKB_KEY_7, XKB_KEY_ampersand,      		TAGKEYS(          XKB_KEY_7, XKB_KEY_ampersand,      
	TAGKEYS(          XKB_KEY_8, XKB_KEY_asterisk,       		TAGKEYS(          XKB_KEY_8, XKB_KEY_asterisk,       
	TAGKEYS(          XKB_KEY_9, XKB_KEY_parenleft,      		TAGKEYS(          XKB_KEY_9, XKB_KEY_parenleft,      
	{ MODKEY|WLR_MODIFIER_ALT,   XKB_KEY_q,           qui |		{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_q,           qui

	/* Ctrl-Alt-Backspace and Ctrl-Alt-Fx used to be hand		/* Ctrl-Alt-Backspace and Ctrl-Alt-Fx used to be hand

config.h
// Taken from https://github.com/djpohly/dwl/issues/466
#define COLOR(hex)    { ((hex >> 24) & 0xFF) / 255.0f, \
                        ((hex >> 16) & 0xFF) / 255.0f, \
                        ((hex >> 8) & 0xFF) / 255.0f, \
                        (hex & 0xFF) / 255.0f }
/* appearance */
static const int sloppyfocus               = 1;  /* focus follows mouse */
static const int bypass_surface_visibility = 0;  /* 1 means idle inhibitors will disable idle tracking even if it's surface isn't visible  */
static const unsigned int borderpx         = 1;  /* border pixel of windows */
static const int showbar                   = 1; /* 0 means no bar */
static const int topbar                    = 1; /* 0 means bottom bar */
static const char *fonts[]                 = {"monospace:size=10"};
static const float rootcolor[]             = COLOR(0x000000ff);
/* vanitygaps start */
static const int smartgaps                 = 0;  /* 1 means no outer gap when there is only one window */
static const int monoclegaps               = 0;  /* 1 means outer gaps in monocle layout */
static const unsigned int gappih           = 10; /* horiz inner gap between windows */
static const unsigned int gappiv           = 10; /* vert inner gap between windows */
static const unsigned int gappoh           = 10; /* horiz outer gap between windows and screen edge */
static const unsigned int gappov           = 10; /* vert outer gap between windows and screen edge */
/* vanitygaps end */
/* This conforms to the xdg-protocol. Set the alpha to zero to restore the old behavior */
static const float fullscreen_bg[]         = {0.0f, 0.0f, 0.0f, 1.0f}; /* You can also use glsl colors */
static uint32_t colors[][3]                = {
	/*               fg          bg          border    */
	[SchemeNorm] = { 0xbbbbbbff, 0x222222ff, 0x444444ff },
	[SchemeSel]  = { 0xeeeeeeff, 0x005577ff, 0x005577ff },
	[SchemeUrg]  = { 0,          0,          0x770000ff },
};

/* tagging */
static char *tags[] = { "1", "2", "3", "4", "5", "6", "7", "8", "9" };

/* logging */
static int log_level = WLR_ERROR;

static const Rule rules[] = {
	/* app_id             title       tags mask     isfloating   monitor */
	{ "Gimp_EXAMPLE",     NULL,       0,            1,           -1 }, /* Start on currently visible tags floating, not tiled */
	{ "firefox_EXAMPLE",  NULL,       1 << 8,       0,           -1 }, /* Start on ONLY tag "9" */
    /* default/example rule: can be changed but cannot be eliminated; at least one rule must exist */
};

/* layout(s) */
static const Layout layouts[] = {
	/* symbol     arrange function */
	{ "[]=",      tile },
	{ "><>",      NULL },    /* no layout function means floating behavior */
	{ "[M]",      monocle },
};

/* monitors */
/* (x=-1, y=-1) is reserved as an "autoconfigure" monitor position indicator
 * WARNING: negative values other than (-1, -1) cause problems with Xwayland clients due to
 * https://gitlab.freedesktop.org/xorg/xserver/-/issues/899 */
static const MonitorRule monrules[] = {
   /* name        mfact  nmaster scale layout       rotate/reflect                x    y
    * example of a HiDPI laptop monitor:
    { "eDP-1",    0.5f,  1,      2,    &layouts[0], WL_OUTPUT_TRANSFORM_NORMAL,   -1,  -1 }, */
    { "HDMI-A-1", 0.5f,  4,      1,    &layouts[0], WL_OUTPUT_TRANSFORM_270,       0,   0},
    { "DP-2",     0.5f,  4,      1,    &layouts[0], WL_OUTPUT_TRANSFORM_NORMAL,    1081,360},
};

/* keyboard */
static const struct xkb_rule_names xkb_rules = {
	/* can specify fields: rules, model, layout, variant, options */
	/* example:
	.options = "ctrl:nocaps",
	*/
	.layout = "br",
	.variant = "abnt2",
	.options = NULL,
};

static const int repeat_rate = 25;
static const int repeat_delay = 600;

/* Start of user-made configuration */

/* Autostart */
static const char *const autostart[] = {
       /* "swaybg", "-o", "HDMI-A-1", "-i", "/home/rafic/Pictures/Wallpapers/cinza-bg.jpg", NULL,
        "swaybg", "-o", "DP-2", "-i", "/home/rafic/Pictures/Wallpapers/montanha-bg.jpg", NULL, */
        "/home/rafic/.config/dwl/autostart.sh", NULL,
        NULL /* terminate */
};

/* End of user-made configuration */

/* Trackpad */
static const int tap_to_click = 1;
static const int tap_and_drag = 1;
static const int drag_lock = 1;
static const int natural_scrolling = 0;
static const int disable_while_typing = 1;
static const int left_handed = 0;
static const int middle_button_emulation = 0;
/* You can choose between:
LIBINPUT_CONFIG_SCROLL_NO_SCROLL
LIBINPUT_CONFIG_SCROLL_2FG
LIBINPUT_CONFIG_SCROLL_EDGE
LIBINPUT_CONFIG_SCROLL_ON_BUTTON_DOWN
*/
static const enum libinput_config_scroll_method scroll_method = LIBINPUT_CONFIG_SCROLL_2FG;

/* You can choose between:
LIBINPUT_CONFIG_CLICK_METHOD_NONE
LIBINPUT_CONFIG_CLICK_METHOD_BUTTON_AREAS
LIBINPUT_CONFIG_CLICK_METHOD_CLICKFINGER
*/
static const enum libinput_config_click_method click_method = LIBINPUT_CONFIG_CLICK_METHOD_BUTTON_AREAS;

/* You can choose between:
LIBINPUT_CONFIG_SEND_EVENTS_ENABLED
LIBINPUT_CONFIG_SEND_EVENTS_DISABLED
LIBINPUT_CONFIG_SEND_EVENTS_DISABLED_ON_EXTERNAL_MOUSE
*/
static const uint32_t send_events_mode = LIBINPUT_CONFIG_SEND_EVENTS_ENABLED;

/* You can choose between:
LIBINPUT_CONFIG_ACCEL_PROFILE_FLAT
LIBINPUT_CONFIG_ACCEL_PROFILE_ADAPTIVE
*/
static const enum libinput_config_accel_profile accel_profile = LIBINPUT_CONFIG_ACCEL_PROFILE_ADAPTIVE;
static const double accel_speed = 0.0;

/* You can choose between:
LIBINPUT_CONFIG_TAP_MAP_LRM -- 1/2/3 finger tap maps to left/right/middle
LIBINPUT_CONFIG_TAP_MAP_LMR -- 1/2/3 finger tap maps to left/middle/right
*/
static const enum libinput_config_tap_button_map button_map = LIBINPUT_CONFIG_TAP_MAP_LRM;

/* If you want to use the windows key for MODKEY, use WLR_MODIFIER_LOGO */
#define MODKEY WLR_MODIFIER_LOGO

#define TAGKEYS(KEY,SKEY,TAG) \
	{ MODKEY,                    KEY,            view,            {.ui = 1 << TAG} }, \
	{ MODKEY|WLR_MODIFIER_CTRL,  KEY,            toggleview,      {.ui = 1 << TAG} }, \
	{ MODKEY|WLR_MODIFIER_SHIFT, SKEY,           tag,             {.ui = 1 << TAG} }, \
	{ MODKEY|WLR_MODIFIER_CTRL|WLR_MODIFIER_SHIFT,SKEY,toggletag, {.ui = 1 << TAG} }

/* helper for spawning shell commands in the pre dwm-5.0 fashion */
#define SHCMD(cmd) { .v = (const char*[]){ "/bin/sh", "-c", cmd, NULL } }

/* commands */
static const char *termcmd[] = { "kitty", NULL };
static const char *menucmd[] = { "fuzzel", NULL };

/* Screenshot command config start */
static const char *screenshot_full[]={"/home/rafic/.local/bin/screenshot.sh", "full", NULL};
static const char *screenshot_area[]={"/home/rafic/.local/bin/screenshot.sh", "area", NULL};
static const char *screenshot_clipboard[]={"/home/rafic/.local/bin/screenshot.sh", "clipboard", NULL};
/* Screenshot command config end */

static const Key keys[] = {
	/* Note that Shift changes certain key codes: 2 -> at, etc. */
	/* modifier                  key                  function          argument */
	{ MODKEY,                    XKB_KEY_d,           spawn,            {.v = menucmd} },
	{ MODKEY, 		     		 XKB_KEY_t,      	  spawn,            {.v = termcmd} },
	{ MODKEY,                    XKB_KEY_b,           togglebar,        {0} },
	{ MODKEY,                    XKB_KEY_j,           focusstack,       {.i = +1} },
	{ MODKEY,                    XKB_KEY_k,           focusstack,       {.i = -1} },
	{ MODKEY,                    XKB_KEY_i,           incnmaster,       {.i = +1} },
	{ MODKEY,                    XKB_KEY_p,           incnmaster,       {.i = -1} },
	{ MODKEY,                    XKB_KEY_h,           setmfact,         {.f = -0.05f} },
	{ MODKEY,                    XKB_KEY_l,           setmfact,         {.f = +0.05f} },
	{ MODKEY,                    XKB_KEY_Return,      zoom,             {0} },
	{ MODKEY,                    XKB_KEY_Tab,         view,             {0} },
	{ MODKEY, 		    		 XKB_KEY_q,           killclient,       {0} },
	{ WLR_MODIFIER_ALT,    		 XKB_KEY_t,           setlayout,        {.v = &layouts[0]} },
	{ WLR_MODIFIER_ALT,    		 XKB_KEY_f,           setlayout,        {.v = &layouts[1]} },
	{ WLR_MODIFIER_ALT,			 XKB_KEY_m,           setlayout,        {.v = &layouts[2]} },
	{ WLR_MODIFIER_ALT,          XKB_KEY_space,       setlayout,        {0} },
	//
	{ WLR_MODIFIER_ALT|WLR_MODIFIER_SHIFT, 		XKB_KEY_space,       togglefloating,   {0} },
	{ WLR_MODIFIER_ALT|WLR_MODIFIER_SHIFT,      XKB_KEY_f,           togglefullscreen, {0} },
	//
	{ MODKEY,                    XKB_KEY_0,           view,             {.ui = ~0} },
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_parenright,  tag,              {.ui = ~0} },
	{ MODKEY,                    XKB_KEY_comma,       focusmon,         {.i = WLR_DIRECTION_LEFT} },
	{ MODKEY,                    XKB_KEY_period,      focusmon,         {.i = WLR_DIRECTION_RIGHT} },
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_less,        tagmon,           {.i = WLR_DIRECTION_LEFT} },
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_greater,     tagmon,           {.i = WLR_DIRECTION_RIGHT} },
	/* vanitygaps start */
	{ MODKEY|WLR_MODIFIER_LOGO,  XKB_KEY_h,          incgaps,       {.i = +1 } },
	{ MODKEY|WLR_MODIFIER_LOGO,  XKB_KEY_l,          incgaps,       {.i = -1 } },
	{ MODKEY|WLR_MODIFIER_LOGO|WLR_MODIFIER_SHIFT,   XKB_KEY_H,      incogaps,      {.i = +1 } },
	{ MODKEY|WLR_MODIFIER_LOGO|WLR_MODIFIER_SHIFT,   XKB_KEY_L,      incogaps,      {.i = -1 } },
	{ MODKEY|WLR_MODIFIER_LOGO|WLR_MODIFIER_CTRL,    XKB_KEY_h,      incigaps,      {.i = +1 } },
	{ MODKEY|WLR_MODIFIER_LOGO|WLR_MODIFIER_CTRL,    XKB_KEY_l,      incigaps,      {.i = -1 } },
	{ MODKEY|WLR_MODIFIER_LOGO,  XKB_KEY_0,          togglegaps,     {0} },
	{ MODKEY|WLR_MODIFIER_LOGO|WLR_MODIFIER_SHIFT,   XKB_KEY_parenright,defaultgaps,    {0} },
	{ MODKEY,                    XKB_KEY_y,          incihgaps,     {.i = +1 } },
	{ MODKEY,                    XKB_KEY_o,          incihgaps,     {.i = -1 } },
	{ MODKEY|WLR_MODIFIER_CTRL,  XKB_KEY_y,          incivgaps,     {.i = +1 } },
	{ MODKEY|WLR_MODIFIER_CTRL,  XKB_KEY_o,          incivgaps,     {.i = -1 } },
	{ MODKEY|WLR_MODIFIER_LOGO,  XKB_KEY_y,          incohgaps,     {.i = +1 } },
	{ MODKEY|WLR_MODIFIER_LOGO,  XKB_KEY_o,          incohgaps,     {.i = -1 } },
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_Y,          incovgaps,     {.i = +1 } },
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_O,          incovgaps,     {.i = -1 } },
	/* vanitygaps end */
	/* rotate-clients start */
	{ WLR_MODIFIER_ALT, 		 XKB_KEY_J,           rotate_clients,   {.i = +1} },
    { WLR_MODIFIER_ALT, 		 XKB_KEY_K,           rotate_clients,   {.i = -1} },
	/* rotate-clients end */
	/* Start of user-made hotkeys */
	{ MODKEY,		     		 XKB_KEY_s,			  spawn,	    	{.v = screenshot_clipboard}},
	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_s,           spawn,            {.v = screenshot_area}},
	{ MODKEY|WLR_MODIFIER_ALT,   XKB_KEY_s,           spawn,            {.v = screenshot_full}},
	/* End of user-made hotkeys */
	TAGKEYS(          XKB_KEY_1, XKB_KEY_exclam,                        0),
	TAGKEYS(          XKB_KEY_2, XKB_KEY_at,                            1),
	TAGKEYS(          XKB_KEY_3, XKB_KEY_numbersign,                    2),
	TAGKEYS(          XKB_KEY_4, XKB_KEY_dollar,                        3),
	TAGKEYS(          XKB_KEY_5, XKB_KEY_percent,                       4),
	TAGKEYS(          XKB_KEY_6, XKB_KEY_asciicircum,                   5),
	TAGKEYS(          XKB_KEY_7, XKB_KEY_ampersand,                     6),
	TAGKEYS(          XKB_KEY_8, XKB_KEY_asterisk,                      7),
	TAGKEYS(          XKB_KEY_9, XKB_KEY_parenleft,                     8),
	{ MODKEY|WLR_MODIFIER_ALT,   XKB_KEY_q,           quit,             {0} },

	/* Ctrl-Alt-Backspace and Ctrl-Alt-Fx used to be handled by X server */
	{ WLR_MODIFIER_CTRL|WLR_MODIFIER_ALT,XKB_KEY_Terminate_Server, quit, {0} },
	/* Ctrl-Alt-Fx is used to switch to another VT, if you don't know what a VT is
	 * do not remove them.
	 */
#define CHVT(n) { WLR_MODIFIER_CTRL|WLR_MODIFIER_ALT,XKB_KEY_XF86Switch_VT_##n, chvt, {.ui = (n)} }
	CHVT(1), CHVT(2), CHVT(3), CHVT(4), CHVT(5), CHVT(6),
	CHVT(7), CHVT(8), CHVT(9), CHVT(10), CHVT(11), CHVT(12),
};

static const Button buttons[] = {
	{ ClkLtSymbol, 0,      BTN_LEFT,   setlayout,      {.v = &layouts[0]} },
	{ ClkLtSymbol, 0,      BTN_RIGHT,  setlayout,      {.v = &layouts[2]} },
	{ ClkTitle,    0,      BTN_MIDDLE, zoom,           {0} },
	{ ClkStatus,   0,      BTN_MIDDLE, spawn,          {.v = termcmd} },
	{ ClkClient,   MODKEY, BTN_LEFT,   moveresize,     {.ui = CurMove} },
	{ ClkClient,   MODKEY, BTN_MIDDLE, togglefloating, {0} },
	{ ClkClient,   MODKEY, BTN_RIGHT,  moveresize,     {.ui = CurResize} },
	{ ClkTagBar,   0,      BTN_LEFT,   view,           {0} },
	{ ClkTagBar,   0,      BTN_RIGHT,  toggleview,     {0} },
	{ ClkTagBar,   MODKEY, BTN_LEFT,   tag,            {0} },
	{ ClkTagBar,   MODKEY, BTN_RIGHT,  toggletag,      {0} },
};
